doc/netplan.md

```diff
@@ -340,6 +340,90 @@ These options are available for all types of interfaces.
      :    Match this policy rule based on the type of service number applied to
           the traffic.
 
+## Authentication
+Netplan supports advanced authentication settings for ethernet and wifi
+interfaces, as well as individual wifi networks, by means of the ``auth`` block.
+
+``auth`` (mapping)
+
+:    Specifies authentication settings for a device of type ``ethernets:`` or
+     ``wifis:``, or a single wifi ``access-points:`` entry. If ``auth`` blocks
+     are specified on both a wifi device and one of its access points, only the
+     settings in the access point's ``auth`` block are taken into account;
+     however, the ``auth`` block of the wifi device, if present, is used for any
+     ``access-points:`` without their own ``auth`` block.
+
+     Example:
+
+          wifis:
+            wlp1s0:
+              auth:
+                key-mgmt: wpa-psk
+                psk: "d3f4ul7"
+              access-points:
+                office:
+                  auth:
+                    # the authentication settings above are ignored for the
+                    # wifi named "office"
+                    key-mgmt: wpa-eap
+                    eap-method: ttls
+                    anonymous-identity: "@internal.example.com"
+                    identity: "joe@internal.example.com"
+                    password: "v4ryS3kr1t"
+                home:
+                  mode: infrastructure
+                  # the authentication settings above are applied to the wifi
+                  # named "home"
+
+      The ``auth`` block supports the following properties:
+
+      ``key-mgmt`` (scalar)
+      :    The supported key management modes are ``none`` (no key management),
+           ``wpa-psk`` (WPA with pre-shared key, common for home wifi),
+           ``wpa-eap`` (WPA with EAP, common for enterprise wifi) and ``8021x``
+           (used primarily for wired Ethernet connections).
+
+      The following properties can be used if ``key-mgmt`` is ``wpa-psk``:
+
+      ``psk`` (scalar)
+      :    The pre-shared key for this connection. Useful only if ``key-mgmt``
+           is ``wpa-psk``.
+
+      ``eap-method`` (scalar)
+      :    The supported EAP methods are ``tls`` (TLS), ``peap`` (Protected EAP)
+           and ``ttls`` (Tunneled TLS). Useful only if ``key-mgmt`` is
+           ``wpa-eap`` or ``8021x``.
+
+      The following properties can be used if ``key-mgmt`` is ``wpa-eap`` or
+      ``8021x``:
+
+      ``identity`` (scalar)
+      :    The identity to use for EAP.
+
+      ``anonymous-identity`` (scalar)
+      :    The identity to pass over the unencrypted channel if the chosen EAP
+           method supports passing a different tunnelled identity.
+
+      ``password`` (scalar)
+      :    The password string for EAP.
```

Comment:
Inconsistent nomenclature.  My understanding is that this is valid for 802.1x as well as for EAP, but the key documentation only mentions EAP.

Which raises another question, my understanding is that 802.1x and EAP are in fact the same protocol, and the only difference is 802.1x is standardized for ethernet.  Why do we support 'wpa-eap' and '8021x' as two separate values for 'key-mgmt' if it's the same protocol?  If these are synonymous, or if the intent is that one should only be used in an ethernet context and the other should only be used in a wifi context, then I think we should collapse them into a single name instead of contributing confusion by having different names for the same thing.

If you consider existing naming so entrenched that saying 'wpa-eap' on an ethernet, or '8021x' on a wifi connection, then I think it should be clear in the documentation that wpa-eap is ONLY supported on wifi (along with wpa-psk), and 8021x is ONLY supported on ethernet.

Reply:
It appears the options dictate the behavior of wpa_supplicant: with `key_mgmt=IEEE8021X`, it performs 802.1X authentication; with `key_mgmt=IEEE8021X` and nonzero `eapol_flags`, it performs 802.1X authentication and requests WEP keys; and with `key_mgmt=WPA-EAP`, it performs 802.1X authentication and then the WPA four-way handshake.

If WEP support is categorically not within the scope of netplan, I can fold both into `eap` and let the value depend on the device type.

Reply:
Thanks, it's helpful to understand the nuance that 802.1x means bare 802.1x without the wpa setup.

I don't want us to paint ourselves into a corner with the key names as a result of ruling out possible use cases.  I also don't want us to make the schema more cumbersome only in the interest of supporting obsolete and insecure protocols that no one would ever reasonably do a new deployment of via netplan.

Are there any valid reasons to use 802.1x either with WEP or with no encryption at all on wifi?

Thinking forward to a possible eventual replacement for wpa2, how would we expect that to be represented?  Would we expect to see an additional set of 'key-mgmt' values for 'wpa3-psk' and 'wpa3-eap'?  If so, I believe it makes sense to keep the current separation between '802.1x' and 'wpa-eap' instead of treating them as synonyms or having context-dependent interpretation.

Reply:
> Are there any valid reasons to use 802.1x either with WEP or with no encryption at all on wifi?

I cannot think of one. It might exist, but it's probably so esoteric that netplan need not support it until someone announces their need on the bug tracker.

> Thinking forward to a possible eventual replacement for wpa2, how would we expect that to be represented? Would we expect to see an additional set of 'key-mgmt' values for 'wpa3-psk' and 'wpa3-eap'?

WPA3 should be published this year; according to [this Aruba blog post](http://community.arubanetworks.com/t5/Technology-Blog/WPA3-The-Next-Generation-in-Secure-Mobility/ba-p/410832), unsecured networks can be supplanted by OWE (RFC8110) and PSK authentication is being superseded by something named *Simultaneous Authentication of Equals*, but all other changes appear to only have an effect on the more specific WPA parameters (baseline rules for EAP; CNSA as a more secure alternative to WPA's TKIP and WPA2's CCMP).

The exhaustive list would thereby probably become `802.1x`, `wpa-psk`, `wpa-eap`, `wpa-owe` and `wpa-sae`. (I wouldn't explicitly mention a WPA version in the values, since most of the methods span multiple versions; e.g. WPA-EAP spans "WPA1" to WPA3.) I bet differentiating between `wpa-psk` and `wpa-sae` will be pretty challenging to users; while wpa_supplicant can probably make use of its "multiple key_mgmt values" feature, NetworkManager only supports one value. I wonder how other Wi-Fi client software will deal with this once WPA3 is released.

There is always the "threat" of a new and completely different security architecture succeeding WPA3, but that could full well get rid of the concept of *key management*, necessitating new netplan configuration keys anyway.

It's pretty obvious that UX was out of scope for the teams designing network authentication, and I don't know how well netplan can counteract that...

Reply:
I think unencrypted wifi isn't an unreasonable thing to support. I can imagine wanting users to know about a network name + sid, wanting to set up custom ip addresses / dns servers, and relying on HTTPS or a VPN.

The reality is that most networks *are* insecure and probably *sniffable*. The WPAx level is just an attempt to prevent someone from stealing the credentials for the network, it isn't a guarantee that no one on the network will sniff. For that, one needs End-to-end encryption (TLS/VPN).

Reply:
Ultimately any network is sniffable, security is relative. There is no absolute security, and we should not rely upon wifi encryption alone, it would be one of several layers. Not supporting WEP means having to have dedicated open SSID's with other protections applied as necessary. Firmware limited to WEP cannot simply be replaced in all cases. Generally it is disappointing wherever a standard is not supported, it renders the whole of product unusable. It is not the job of the framework to make these decisions.